README.md
UAC
===

User-Agent Classifier: utilities and libraries.

Compile and run test
====================

 * cmake
 * make
 * run ./uac_detect to test

```console
# echo "Mozilla/5.0 (Linux; Android 4.4.2; AppleWebKit/537.36 (KHTML, like Gecko) Chrome/31.0.1650.59 Safari/537.36" |  ./uac_detect .
```



debian/changelog
uac (0.2.0-0) unstable; urgency=high

  * New version of UAC library and tools

 -- Alexander Pankov <pianist@usrsrc.ru>  Tue, 13 Jan 2015 06:24:00 +0300



src/sla.c
/*
 * Copyright (C) 2007, libturglem development team.
 *
 * This file is released under the LGPL.
*/

#include "sla.h"
#include <string.h>

static int sla_convert_array_initiated = 0;
static MAFSA_letter sla_convert_array[256] = { 0 };

static void sla_convert_array_init()
{
	int i;
	memset(sla_convert_array, 10, 256);

	for (i = 0; i <= 9; ++i)
	{
		sla_convert_array['0' + i] = i;
	}

	sla_convert_array['-'] = 11;
	sla_convert_array['|'] = MAX_LETTER_SLA;

	for (i = 0; i < 26; ++i)
	{
		sla_convert_array['a' + i] = i + 12;
		sla_convert_array['A' + i] = i + 12;
	}

	sla_convert_array_initiated = 1;
}

ssize_t conv_s2l_sla(const char *s, MAFSA_letter *l, size_t sz)
{
	if (!sla_convert_array_initiated)
	{
		sla_convert_array_init();
	}

	ssize_t pos = 0;
	const char *i = s;
	while (*i && (*i != '\n') && (pos < sz))
	{
		MAFSA_letter _n = sla_convert_array[(unsigned char)*i++];

		if (255 == _n) continue;

		if ((10 == _n) && pos)
		{
			if (10 == l[pos-1]) continue;
		}

		l[pos++] = _n;
	}

	return pos;
}

size_t conv_l2s_sla(const MAFSA_letter *l, size_t sz_l, char *s, size_t sz_s)
{
	size_t pos_w = 0;
	size_t pos_r = 0;

	while ((pos_w + 1 < sz_s) && (pos_r < sz_l))
	{
		if(l[pos_r] < 10)
		{
			s[pos_w++] = l[pos_r] + '0';
		}
		else if ((l[pos_r] < 38) && (l[pos_r] > 11))
		{
			s[pos_w++] = l[pos_r] - 12 + 'A';
		}
		else
		{
			switch (l[pos_r])
			{
				case 10:
					s[pos_w++] = '_';
					break;
				case 11:
					s[pos_w++] = '-';
					break;
				default:
					s[pos_w++] = '|';
					break;
			}
		}
		pos_r++;
	}
	s[pos_w] = 0;
	return pos_w;
}

ssize_t conv_b2l_sla(const char *s, size_t sz_s, MAFSA_letter *l, size_t sz_l)
{
	ssize_t pos = 0;
	const char *i = s;
	const char *e = s + sz_s;
	while ((i < e) && (*i != '\n') && (pos < sz_l))
	{
		if (i[0] >= '0' && i[0] <= '9')
		{
			l[pos++] = i[0] - '0';
			i++;
		}
		else if (strchr(" \t_@./+\\\"'=%$#@!?(){}[]", i[0]))
		{
			if (pos && (l[pos-1] != 10) && (l[pos-1] != 11)) l[pos++] = 10;
			i++;
		}
		else if (strchr("-!#$%^&*().\\", i[0]))
		{
			if (pos && (l[pos-1] != 10) && (l[pos-1] != 11)) l[pos++] = 11; /* '-' */
			i++;
		}
		else if (i[0] >= 'a' && i[0] <= 'z')
		{
			l[pos++] = i[0] - 'a' + 12;
			i++;
		}
		else if (i[0] >= 'A' && i[0] <= 'Z')
		{
			l[pos++] = i[0] - 'A' + 12;
			i++;
		}
		else if (i[0] == '|')
		{
			l[pos++] = MAX_LETTER_SLA;
			i++;
		}
		else i++;
	}
	return pos;
}

size_t conv_l2b_sla(const MAFSA_letter *l, size_t sz_l, char *s, size_t sz_s)
{
	size_t pos_w = 0;
	size_t pos_r = 0;

	while ((pos_w < sz_s) && (pos_r < sz_l))
	{
		if(l[pos_r] < 10)
		{
			s[pos_w++] = l[pos_r] + '0';
		}
		else if ((l[pos_r] < 38) && (l[pos_r] > 11))
		{
			s[pos_w++] = l[pos_r] - 12 + 'A';
		}
		else if (pos_w)
		{
			switch (l[pos_r])
			{
				case 10:
					s[pos_w++] = '_';
					break;
				case 11:
					s[pos_w++] = '-';
					break;
				default:
					s[pos_w++] = '|';
					break;
			}
		}
		pos_r++;
	}
	return pos_w;
}



src/uac.c
#include "uac.h"
#include "sla.h"
#include <errno.h>
#include <string.h>
#include <stdio.h>
#include <MAFSA/automaton.h>

#define UAC_DEFINE_GROUP(name) { UAC_GROUP_ ## name, "g_" # name, 0 },
#define UAC_DEFINE_FLAG(name,_groups) { UAC_FLAG_ ## name, # name, _groups },

#define UAC_GROUP_MASK_ANY	((uint64_t)0xFFFFFFFFFFFFFFFF)
#define UAC_GROUP_MASK_other	((uint64_t)1 << UAC_GROUP_other)
#define UAC_GROUP_MASK_android	((uint64_t)1 << UAC_GROUP_android)
#define UAC_GROUP_MASK_windows	((uint64_t)1 << UAC_GROUP_windows)
#define UAC_GROUP_MASK_apple	((uint64_t)1 << UAC_GROUP_apple)
#define UAC_GROUP_MASK_unix	((uint64_t)1 << UAC_GROUP_unix)

uac_flag_info_t uac_flag_info[] = {
	UAC_DEFINE_GROUP(other)
	UAC_DEFINE_GROUP(android)
	UAC_DEFINE_GROUP(apple)
	UAC_DEFINE_GROUP(windows)
	UAC_DEFINE_GROUP(unix)

	UAC_DEFINE_FLAG(phone,	UAC_GROUP_MASK_ANY)
	UAC_DEFINE_FLAG(tablet,	UAC_GROUP_MASK_ANY)
	UAC_DEFINE_FLAG(tv,	UAC_GROUP_MASK_ANY)

	UAC_DEFINE_FLAG(64bit,	UAC_GROUP_MASK_ANY)
	UAC_DEFINE_FLAG(arm,	UAC_GROUP_MASK_ANY)

	UAC_DEFINE_FLAG(android,	UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(android_old,	UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(android_2_3_x,	UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(android_3_x,	UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(android_4,	UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(android_api_16,	UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(android_api_17,	UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(android_api_18,	UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(android_api_19,	UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(android_api_20,	UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(android_5,	UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(android_api_21,	UAC_GROUP_MASK_android)

	UAC_DEFINE_FLAG(linux,	UAC_GROUP_MASK_unix)
	UAC_DEFINE_FLAG(bsd,	UAC_GROUP_MASK_unix)
	UAC_DEFINE_FLAG(cros,	UAC_GROUP_MASK_unix)

	UAC_DEFINE_FLAG(debian,	UAC_GROUP_MASK_unix)
	UAC_DEFINE_FLAG(ubuntu,	UAC_GROUP_MASK_unix)
	UAC_DEFINE_FLAG(gentoo,	UAC_GROUP_MASK_unix)

	UAC_DEFINE_FLAG(ipod,	UAC_GROUP_MASK_apple)
	UAC_DEFINE_FLAG(iphone,	UAC_GROUP_MASK_apple)
	UAC_DEFINE_FLAG(ipad,	UAC_GROUP_MASK_apple)
	UAC_DEFINE_FLAG(mac,	UAC_GROUP_MASK_apple)

	UAC_DEFINE_FLAG(win_old,	UAC_GROUP_MASK_windows)
	UAC_DEFINE_FLAG(win_xp,		UAC_GROUP_MASK_windows)
	UAC_DEFINE_FLAG(win_vista,	UAC_GROUP_MASK_windows)
	UAC_DEFINE_FLAG(win_7,		UAC_GROUP_MASK_windows)
	UAC_DEFINE_FLAG(win_8,		UAC_GROUP_MASK_windows)
	UAC_DEFINE_FLAG(win_10,		UAC_GROUP_MASK_windows)

	UAC_DEFINE_FLAG(symbian,	UAC_GROUP_MASK_other)
	UAC_DEFINE_FLAG(nokia,		UAC_GROUP_MASK_other + UAC_GROUP_MASK_windows)
	UAC_DEFINE_FLAG(blackberry,	UAC_GROUP_MASK_other)
	UAC_DEFINE_FLAG(samsung,	UAC_GROUP_MASK_other + UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(nexus,		UAC_GROUP_MASK_other + UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(huawei,		UAC_GROUP_MASK_other + UAC_GROUP_MASK_android)
	UAC_DEFINE_FLAG(htc,		UAC_GROUP_MASK_other + UAC_GROUP_MASK_android + UAC_GROUP_MASK_windows)
	UAC_DEFINE_FLAG(lenovo,		UAC_GROUP_MASK_other + UAC_GROUP_MASK_android + UAC_GROUP_MASK_windows)

	UAC_DEFINE_FLAG(chrome,		UAC_GROUP_MASK_ANY)
	UAC_DEFINE_FLAG(firefox,	UAC_GROUP_MASK_ANY)
	UAC_DEFINE_FLAG(safari,		UAC_GROUP_MASK_ANY)
	UAC_DEFINE_FLAG(ie,		UAC_GROUP_MASK_windows)
	UAC_DEFINE_FLAG(gnome,		UAC_GROUP_MASK_other + UAC_GROUP_MASK_unix)
	UAC_DEFINE_FLAG(opera,		UAC_GROUP_MASK_ANY)
	UAC_DEFINE_FLAG(yabrowser,	UAC_GROUP_MASK_ANY)
	UAC_DEFINE_FLAG(corom,		UAC_GROUP_MASK_ANY)
	UAC_DEFINE_FLAG(maxthon,	UAC_GROUP_MASK_ANY)
	UAC_DEFINE_FLAG(konqueror,	UAC_GROUP_MASK_other + UAC_GROUP_MASK_unix)
	UAC_DEFINE_FLAG(links,		UAC_GROUP_MASK_other + UAC_GROUP_MASK_unix)

	UAC_DEFINE_FLAG(ie_old,		UAC_GROUP_MASK_windows)
	UAC_DEFINE_FLAG(ie_latest,	UAC_GROUP_MASK_windows)
	UAC_DEFINE_FLAG(ie_11,		UAC_GROUP_MASK_windows)
	UAC_DEFINE_FLAG(ie_10,		UAC_GROUP_MASK_windows)

	UAC_DEFINE_FLAG(maxthon_cloud,	UAC_GROUP_MASK_ANY)
	{ 0, NULL, 0 }
};

static MAFSA_automaton uac_ma_base = 0;

inline static void fill_error(const char* fname, char* err_buf, unsigned err_buf_sz)
{
	if (err_buf)
	{
		snprintf(err_buf, err_buf_sz, "Can't load %s: (%d) %s", fname, errno, strerror(errno));
	}
}

int uac_init(const char *dict_dir, char* err_buf, unsigned err_buf_sz)
{
	char fname[1024];

	snprintf(fname, 1024, "%s/uac.automaton", dict_dir);
	uac_ma_base = MAFSA_automaton_load_from_binary_file(fname, 0);

	if (!uac_ma_base)
	{
		fill_error(fname, err_buf, err_buf_sz);
		return -1;
	}

	return 0;
}

void uac_free()
{
	MAFSA_automaton_close(uac_ma_base);
}

#define MAX_MAFSA_LEN 1024

struct uac_enum_working_s
{
	uint64_t flags_plus[UAC_GROUP_MAX + 1];
	uint64_t flags_minus[UAC_GROUP_MAX + 1];
};

static void MAFSACALL uac_determine_callback(void* user_data, const MAFSA_letter* l, size_t sz)
{
	if (!user_data) return;

	struct uac_enum_working_s* ws = (struct uac_enum_working_s*)user_data;

	size_t i = 0;
	while (i < sz)
	{
		if (MAX_LETTER_SLA == l[i++]) break;
	}

	uint32_t group_id = 0;
	uint32_t mul = 1;
	while ((MAX_LETTER_SLA != l[i]) && (i < sz))
	{
		group_id += mul * l[i++];
		mul *= MAX_LETTER_SLA;
	}

	if (1 == mul) group_id = UAC_GROUP_MAX;

	i++;
	uint32_t flag_id = 0;
	mul = 1;
	while ((MAX_LETTER_SLA != l[i]) && (i < sz))
	{
		flag_id += mul * l[i++];
		mul *= MAX_LETTER_SLA;
	}

	/* this should not happen if automaton is correct */
	if (group_id > UAC_GROUP_MAX) return;

	i++;
	if (1 != mul)
	{
		if (i < sz)// LETTER_MINUS at the end
		{
			ws->flags_minus[group_id] |= ((uint64_t)1 << flag_id);
		}
		else
		{
			ws->flags_plus[group_id] |= ((uint64_t)1 << flag_id);
		}
	}
}

static int uac_do_classify(uac_result_t* ret, const MAFSA_letter* l, ssize_t ssz)
{
	MAFSA_letter tmp[MAX_MAFSA_LEN];

	struct uac_enum_working_s ws;
	memset(&ws, 0, sizeof(struct uac_enum_working_s));

	MAFSA_automaton_search_enumerate(uac_ma_base, l, ssz, tmp, MAX_MAFSA_LEN, &ws, MAX_LETTER_SLA, uac_determine_callback);

	uint64_t base_flags = (ws.flags_plus[UAC_GROUP_MAX] & ~ws.flags_minus[UAC_GROUP_MAX]);

	uint32_t group_id = 0;
	uint64_t it_flag = 1;
	while (it_flag)
	{
		if ((base_flags & it_flag) == base_flags) break;

		group_id++;
		it_flag <<= 1;
	}

	ret->group_id = group_id;
	ret->flags = 0;

	uint64_t group_flags = (ws.flags_plus[group_id] & ~ws.flags_minus[group_id]);
	ret->flags = group_flags;

	return 0;
}

int uac_classify(uac_result_t* ret, const char* s)
{
	MAFSA_letter l[MAX_MAFSA_LEN];
	ssize_t ssz = conv_s2l_sla(s, l, 1024);

	if (ssz <= 0) return -1;

	return uac_do_classify(ret, l, ssz);
}

int uac_classifyi_binary(uac_result_t* ret, const char* s, size_t sz)
{
	MAFSA_letter l[MAX_MAFSA_LEN];
	ssize_t ssz = conv_b2l_sla(s, sz, l, 1024);

	if (ssz <= 0) return -1;

	return uac_do_classify(ret, l, ssz);
}

static int uac_bit_text_initiated = 0;

#define UAC_GROUP_COUNT			(UAC_GROUP_unix + 1)
#define UAC_FLAG_BITS_COUNT		64
#define UAC_MAX_FLAG_TITLE_STRLEN	32
static const char* uac_bit_texts[UAC_GROUP_COUNT][UAC_FLAG_BITS_COUNT] = { { 0 } };
static const char* uac_group_titles[UAC_GROUP_COUNT] = { 0 };

static void uac_bit_text_init()
{
	int i;
	int j;

	for (i = 0; uac_flag_info[i].title; ++i)
	{
		if (!uac_flag_info[i].groups)
		{
			uac_group_titles[uac_flag_info[i].flag_id] = uac_flag_info[i].title;
		}
		else
		{
			for (j = 0; j < UAC_GROUP_COUNT; ++j)
			{
				if ((uac_flag_info[i].groups & ((uint64_t)1 << j)))
				{
					uac_bit_texts[j][uac_flag_info[i].flag_id] = uac_flag_info[i].title;
				}
			}
		}
	}

	uac_bit_text_initiated = 1;
}

#define WRITE_STR(_s) \
{ \
	*ptr = 0; \
	if (_s) \
	{ \
		if (strlen(_s) + 1 > (sz - (ptr - buf))) return; \
		ptr = stpcpy(ptr,_s); \
	} \
}

void uac_hr_cl_result(uac_result_t* ret, char* buf, size_t sz)
{
	if (!uac_bit_text_initiated)
	{
		uac_bit_text_init();
	}

	if (!sz) return;

	char* ptr = buf;
	if (ret->group_id >= UAC_GROUP_COUNT)
	{
		WRITE_STR("unknown");
		return;
	}
	WRITE_STR(uac_group_titles[ret->group_id]);
	WRITE_STR(":");

	int ft = 1;
	int i;
	for (i = 0; i < UAC_FLAG_BITS_COUNT; ++i)
	{
		if ((ret->flags & ((uint64_t)1 << i)))
		{
			if (!ft) WRITE_STR(",");
			ft = 0;

			WRITE_STR(uac_bit_texts[ret->group_id][i]);
		}
	}
}



src/uacc.cpp
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <string>
#include <map>
#include <MAFSA/daciuk.hpp>
#include "sla.h"
#include "uac.h"

typedef std::map<std::string, uint32_t> uac_std_flag_title_rev_t;
typedef MAFSA::daciuk<MAX_LETTER_SLA + 1> dict_compiler_t;

uac_std_flag_title_rev_t flag_title_rev;

void process_input_file(dict_compiler_t& dc, const char* fname, int param_no)
{
	FILE* f = fopen(fname, "r");

	if (!f)
	{
		fprintf(stderr, "Can't open %s for reading: %s\n", fname, strerror(errno));
		exit(-1);
	}

	unsigned line_no = 1;
	while (!feof(f))
	{
		char buf[1024];
		if (!fgets(buf, 1024, f)) break;

		if ('#' == buf[0]) continue;

		char* delim = strpbrk(buf, " \t:");
		if (!delim) continue;
		
		*delim++ = 0;
		//while (strchr(delimiters, delim)) delim++;

		char* nlrl = strpbrk(delim, "\r\n|");
		if (nlrl)
		{
			line_no++;
			*nlrl = 0;
		}

		MAFSA_letter l[1024];
		ssize_t ssz = conv_s2l_sla(delim, l, 1024);
		if (ssz > 0)
		{
			l[ssz++] = MAX_LETTER_SLA;

			char* tk = buf;
			while (tk)
			{
				char* ntk = strpbrk(tk, ",;");
				if (ntk) *ntk++ = 0;

				ssize_t len = ssz;
#ifdef _DEBUG
ssize_t save_len = len;
#endif
				/* write dictionary id if not group_id */
				if (param_no >= 0)
				{
					int _p = param_no;
					do
					{
						l[len++] = _p % MAX_LETTER_SLA;
						_p /= MAX_LETTER_SLA;
					}
					while (_p);
				}

				l[len++] = MAX_LETTER_SLA;

				const char* to_search = tk;
				if ('-' == *to_search) to_search++;

				uac_std_flag_title_rev_t::const_iterator it = flag_title_rev.find(to_search);
				if (it != flag_title_rev.end())
				{
					uint32_t id = it->second;
					do
					{
						l[len++] = id % MAX_LETTER_SLA;
						id /= MAX_LETTER_SLA;
					}
					while (id);
				}
				else
				{
					fprintf(stderr, "Unknown flag '%s', at line %u in %s can't continue...\n", to_search, line_no, fname);
					exit(-1);
				}

				l[len++] = MAX_LETTER_SLA;

				if ('-' == *tk) l[len++] = LETTER_MINUS;

				dc.insert(l, len);

#ifdef _DEBUG
char x[1024] = {0};
conv_l2s_sla(l, len, x, 1024);
printf("%s -- %s\n", x, buf);
uint32_t id = 0;
uint32_t mul = 1;
while (l[save_len] != MAX_LETTER_SLA) { id += mul * l[save_len++]; mul *= MAX_LETTER_SLA;}
if (mul == 1) { printf("\tno id\n"); } else { printf("\tid %u\n", id); }
#endif

				tk = ntk;
			}
		}
	}

	fclose(f);
}

int main(int argc, char **argv)
{
	for (int i = 0; uac_flag_info[i].title; ++i)
	{
		flag_title_rev[uac_flag_info[i].title] = uac_flag_info[i].flag_id;
	}

        if (argc != 3)
        {
                printf("usage: uacc uac.automaton /path/to/dict/dir\n");
                return -1;
        }

	dict_compiler_t dc;

	char fname[1024];
	snprintf(fname, 1024, "%s/base.txt", argv[2]);
	process_input_file(dc, fname, -1);

	for (int i = 0; uac_flag_info[i].title; ++i)
	{
		if (!uac_flag_info[i].groups)
		{
			snprintf(fname, 1024, "%s/%s.txt", argv[2], uac_flag_info[i].title);
			process_input_file(dc, fname, uac_flag_info[i].flag_id);
		}
	}

	dc.save_to_file(argv[1]);
	return 0;
}



src/uac.h
#ifndef __UAC_H__USER_AGENT_CLASSIFIER__
#define __UAC_H__USER_AGENT_CLASSIFIER__

#include <stdint.h>
#include <stddef.h>

#ifdef __cplusplus
extern "C" {
#endif

#define UAC_GROUP_other		0
#define UAC_GROUP_android	1
#define UAC_GROUP_windows	2
#define UAC_GROUP_apple		3
#define UAC_GROUP_unix		4

#define UAC_GROUP_unknown	64
#define UAC_GROUP_MAX		UAC_GROUP_unknown

/* Flags for groups

   Other
   DDD.DD.. ........ ........ MMMMMMMM ........ BBBBBBBB BB...... ....b...

   Android
   DDD.DD.A aaaaaaaa aa...... .M.MMMMM ........ BBB.BBBB ........ ....b...

   Windows
   DDD.DD.. wwwwww.. ........ ......MM ........ BBBBBBBB ........ bbbbb...

   Apple
   DDD.DD.. dddd.... ........ ........ ........ BBB.BBBB ........ b...b...
   D: device type
   d: device kind

   Unix
   DDD.DD.. SSS..... ddd..... ........ ........ BBBBBBBB BB...... ....b...

   D: device type
   A: set iff Android API version detected
   S: system type
   d: distro
   a: Android version info bits
   w: Windows version info bits
   M: Manufactor, brand info bits
   B: Browser type info bits
   b: Browser versions info bits
   .: unused

*/

/* common flags */
#define UAC_FLAG_phone		0
#define UAC_FLAG_tablet		1
#define UAC_FLAG_tv		2

#define UAC_MASK_device		((1 << UAC_FLAG_phone) + (1 << UAC_FLAG_tablet) + (1 << UAC_FLAG_tv))

#define UAC_FLAG_64bit		4
#define UAC_FLAG_arm		5

/* android API versions */
/* UAC_FLAG_android == API version detected */
#define UAC_FLAG_android		7
/* android_less_4 == all versions before Android 4.0 (API 1 ... 13) */
#define UAC_FLAG_android_old		8
#define UAC_FLAG_android_2_3_x		9
#define UAC_FLAG_android_3_x		10
/* android_4 == all Android 4.x versions (API 14 ... 20) */
#define UAC_FLAG_android_4		11
#define UAC_FLAG_android_api_16		12
#define UAC_FLAG_android_api_17		13
#define UAC_FLAG_android_api_18		14
#define UAC_FLAG_android_api_19		15
#define UAC_FLAG_android_api_20		16
/* android_5 == all Android 5.x versions (API 21 ...) */
#define UAC_FLAG_android_5		17
#define UAC_FLAG_android_api_21		18
#define UAC_MASK_android_version	(((uint64_t)1 << UAC_FLAG_android_old) + ((uint64_t)1 << UAC_FLAG_android_2_3_x) + ((uint64_t)1 << UAC_FLAG_android_3_x) + ((uint64_t)1 << UAC_FLAG_android_4) + ((uint64_t)1 << UAC_FLAG_android_api_16) + ((uint64_t)1 << UAC_FLAG_android_api_17) + ((uint64_t)1 << UAC_FLAG_android_api_18) + ((uint64_t)1 << UAC_FLAG_android_api_19) + ((uint64_t)1 << UAC_FLAG_android_api_20) + ((uint64_t)1 << UAC_FLAG_android_5) + ((uint64_t)1 << UAC_FLAG_android_api_21))

/* windows versions */
#define UAC_FLAG_win_old	8
#define UAC_FLAG_win_xp		9
#define UAC_FLAG_win_vista	10
#define UAC_FLAG_win_7		11
#define UAC_FLAG_win_8		12
#define UAC_FLAG_win_10		13
#define UAC_MASK_win_os_ver	(((uint64_t)1 << UAC_FLAG_win_old) + ((uint64_t)1 << UAC_FLAG_win_xp) + ((uint64_t)1 << UAC_FLAG_win_vista) + ((uint64_t)1 << UAC_FLAG_win_7) + ((uint64_t)1 << UAC_FLAG_win_8) + ((uint64_t)1 << UAC_FLAG_win_10))

/* unix oses (x11 or console based) */
#define UAC_FLAG_linux		8
#define UAC_FLAG_bsd		9
#define UAC_FLAG_cros		10

/* linux distros */
#define UAC_FLAG_debian		16
#define UAC_FLAG_ubuntu		17
#define UAC_FLAG_gentoo		18

/* Apple */
#define UAC_FLAG_ipod		8
#define UAC_FLAG_iphone		9
#define UAC_FLAG_ipad		10
#define UAC_FLAG_mac		11

/* mobiles (android and not) */
#define UAC_FLAG_symbian	24
#define UAC_FLAG_nokia		25
#define UAC_FLAG_blackberry	26
#define UAC_FLAG_samsung	27
#define UAC_FLAG_nexus		28
#define UAC_FLAG_huawei		29
#define UAC_FLAG_htc		30
#define UAC_FLAG_lenovo		31

/* browsers */
#define UAC_FLAG_chrome		40
#define UAC_FLAG_firefox	41
#define UAC_FLAG_safari		42
/* IE and Gnome (Galeon,Epiphany,Web...) have the same flag id, because IE is win browser, and Gnome browsers are X11 */
#define UAC_FLAG_ie		43
#define UAC_FLAG_gnome		43
#define UAC_FLAG_opera		44
#define UAC_FLAG_yabrowser	45
#define UAC_FLAG_corom		46
#define UAC_FLAG_maxthon	47
#define UAC_FLAG_konqueror	48
#define UAC_FLAG_links		49
#define UAC_MASK_browsers	(((uint64_t)1 << UAC_FLAG_chrome) + ((uint64_t)1 << UAC_FLAG_firefox) + ((uint64_t)1 << UAC_FLAG_safari) + ((uint64_t)1 << UAC_FLAG_ie) + ((uint64_t)1 << UAC_FLAG_opera) + ((uint64_t)1 << UAC_FLAG_yabrowser) + ((uint64_t)1 << UAC_FLAG_corom) + ((uint64_t)1 << UAC_FLAG_maxthon) + ((uint64_t)1 << UAC_FLAG_konqueror) + ((uint64_t)1 << UAC_FLAG_links))

/* ie flags */
#define UAC_FLAG_ie_latest	56
#define UAC_FLAG_ie_old		57
#define UAC_FLAG_ie_11		58
#define UAC_FLAG_ie_10		59

/* Maxthon browser flags */
#define UAC_FLAG_maxthon_cloud	60



int uac_init(const char *dict_dir, char* err_buf, unsigned err_buf_sz);
void uac_free();

struct uac_result_s
{
	uint32_t group_id;
	uint64_t flags;
};
typedef struct uac_result_s uac_result_t;

int uac_classify(uac_result_t* ret, const char* s);
int uac_classifyi_binary(uac_result_t* ret, const char* s, size_t sz);

void uac_hr_cl_result(uac_result_t* ret, char* buf, size_t sz);

/* flags info */

struct uac_flag_info_s
{
	uint32_t flag_id;
	const char* title;
	uint64_t groups;
};

typedef struct uac_flag_info_s uac_flag_info_t;
extern uac_flag_info_t uac_flag_info[];

#ifdef __cplusplus
}
#endif

#endif /* __UAC_H__USER_AGENT_CLASSIFIER__ */
